code/game/mecha/mech_fabricator.dm

```diff
@@ -281,151 +320,103 @@
-				part_set = null
-			else
-				part_set = tpart_set
-				screen = "parts"
-	if(href_list["part"])
-		var/T = afilter.getStr("part")
-		for(var/v in files.known_designs)
-			var/datum/design/D = files.known_designs[v]
+
+	if(href_list["menu"]) // Main menu
+		screen = MECHFAB_MAIN_MENU
+
+	if(href_list["category"]) // Inside a "category": cyborg parts, medical etc.
+		selected_category = href_list["category"]
+		screen = MECHFAB_CATEGORY_MENU
+
+	if(href_list["search"])
+		if(href_list["to_search"])
+			temp_search = href_list["to_search"]
+		if(!temp_search)
+			return
+		matching_designs.Cut()
+
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
 			if(D.build_type & fabricator_type)
-				if(D.id == T)
-					if(!processing_queue)
-						build_part(D)
-					else
-						add_to_queue(D)
-					break
+				if(findtext(D.name, temp_search))
+					matching_designs.Add(D)
+		screen = MECHFAB_SEARCH_MENU
+
 	if(href_list["add_to_queue"])
-		var/T = afilter.getStr("add_to_queue")
-		for(var/v in files.known_designs)
-			var/datum/design/D = files.known_designs[v]
+		var/design = href_list["add_to_queue"]
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
 			if(D.build_type & fabricator_type)
-				if(D.id == T)
+				if(D.id == design)
 					add_to_queue(D)
 					break
-		return update_queue_on_page()
+
+	if(href_list["category_add_all"])
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
+			if(D.build_type & fabricator_type && selected_category in D.category)
+				add_to_queue(D)
+
+	if(href_list["main_menu_add_all"])
+		selected_category = href_list["main_menu_add_all"]
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
+			if(D.build_type & fabricator_type && selected_category in D.category)
+				add_to_queue(D)
+		selected_category = null
+
 	if(href_list["remove_from_queue"])
-		remove_from_queue(afilter.getNum("remove_from_queue"))
-		return update_queue_on_page()
-	if(href_list["partset_to_queue"])
-		add_part_set_to_queue(afilter.get("partset_to_queue"))
-		return update_queue_on_page()
+		var/index = text2num(href_list["remove_from_queue"])
+		remove_from_queue(index)
+
 	if(href_list["process_queue"])
-		spawn(0)
-			if(processing_queue || being_built)
-				return FALSE
-			processing_queue = 1
-			process_queue()
-			processing_queue = 0
-	if(href_list["clear_temp"])
-		temp = null
-	if(href_list["screen"])
-		screen = href_list["screen"]
-	if(href_list["queue_move"] && href_list["index"])
-		var/index = afilter.getNum("index")
-		var/new_index = index + afilter.getNum("queue_move")
-		if(isnum(index) && isnum(new_index))
-			if(IsInRange(new_index,1,queue.len))
-				queue.Swap(index,new_index)
-		return update_queue_on_page()
+		if(processing_queue || being_built)
+			return FALSE
+		processing_queue = TRUE
+		process_queue()
+		processing_queue = FALSE
+
 	if(href_list["clear_queue"])
 		queue = list()

```

Comment:
Not your code, but while you're touching it:
use `queue.Cut()` instead.
Subtle difference, and unlikely to be relevant in practice, but `.Cut()` ensures that if we had `process_queue(src.queue)`, then that, passed `queue` gets cleared as well.